File: python/fourty/fourty.py
def good_pairs(nums):
    from collections import defaultdict
    counts = defaultdict(int)

    for num in nums:
        counts[num] += 1

    pairs = 0
    for key in counts.keys():
        if counts[key] > 1:
            pairs += int(counts[key] * (counts[key] - 1) / 2)
    return pairs

def unique_occurrences(nums):
    from collections import defaultdict
    counts = defaultdict(int)

    for num in nums:
        counts[num] += 1

    unique_counts = set()
    for val in counts.values():
        if val in unique_counts:
            return False
        unique_counts.add(val)
    return True

def reverse_ll(head):
    prev, curr = None, head

    while curr:
        nxt = curr.next
        curr.next = prev
        prev = curr
        curr = nxt
    return prev

class ListNode():
    def __init__(self, val=None, nxt=None):
        self.val = val
        self.next = nxt

def merge_lists(list1, list2):
    dummy = ListNode()
    helper = dummy

    while list1 or list2:
        if list1 is None:
            helper.next = ListNode(list2.val)
            helper, list2 = helper.next, list2.next
        elif list2 is None:
            helper.next = ListNode(list1.val)
            helper, list1 = helper.next, list1.next
        else:
            if list1.val <= list2.val:
                helper.next = ListNode(list1.val)
                helper, list1 = helper.next, list1.next
            else:
                helper.next = ListNode(list2.val)
                helper, list2 = helper.next, list2.next
    return dummy.next

File: go/fourty/fourty.go
package fourty

func IdenticalPairs(nums []int) int {
	counts := make(map[int]int)

	for i := range nums {
		counts[nums[i]] += 1
	}

	pairs := 0
	for _, val := range counts {
		if val > 1 {
			pairs += val * (val - 1) / 2
		}
	}
	return pairs
}

func UniqueOccurrences(nums []int) bool {
	counts := make(map[int]int)

	for i := range nums {
		counts[nums[i]] += 1
	}

	uniqueCounts := make(map[int]int)
	for _, val := range counts {
		if _, exists := uniqueCounts[val]; exists {
			return false
		} else {
			uniqueCounts[val] = val
		}
	}
	return true
}

type ListNode struct {
	Val  int
	Next *ListNode
}

func ReverseLL(head *ListNode) *ListNode {
	var prev *ListNode = nil
	curr := head
	for curr != nil {
		nxt := curr.Next
		curr.Next = prev
		prev = curr
		curr = nxt
	}
	return prev
}

func MergeTwoLists(list1, list2 *ListNode) *ListNode {
	dummy := &ListNode{}
	helper := dummy
	for list1 != nil || list2 != nil {
		if list1 == nil {
			helper.Next = &ListNode{list2.Val, nil}
			helper = helper.Next
			list2 = list2.Next
		} else if list2 == nil {
			helper.Next = &ListNode{list1.Val, nil}
			helper = helper.Next
			list1 = list1.Next
		} else {
			if list1.Val <= list2.Val {
				helper.Next = &ListNode{list1.Val, nil}
				helper = helper.Next
				list1 = list1.Next
			} else {
				helper.Next = &ListNode{list2.Val, nil}
				helper = helper.Next
				list2 = list2.Next
			}
		}
	}
	return dummy.Next
}
